src/jit/lsraxarch.cpp

```diff
@@ -2497,32 +2497,71 @@ void LinearScan::TreeNodeInfoInitSIMD(GenTreeSIMD* simdTree, TreeNodeInfo* info)
 void LinearScan::TreeNodeInfoInitHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree, TreeNodeInfo* info)
 {
     NamedIntrinsic intrinsicID = intrinsicTree->gtHWIntrinsicId;
-    InstructionSet isa         = compiler->isaOfHWIntrinsic(intrinsicID);
+    InstructionSet isa         = Compiler::isaOfHWIntrinsic(intrinsicID);
     if (isa == InstructionSet_AVX || isa == InstructionSet_AVX2)
     {
         SetContainsAVXFlags(true, 32);
     }
-    info->srcCount += GetOperandInfo(intrinsicTree->gtOp.gtOp1);
-    if (intrinsicTree->gtGetOp2IfPresent() != nullptr)
+    GenTree* op1   = intrinsicTree->gtOp.gtOp1;
+    GenTree* op2   = intrinsicTree->gtOp.gtOp2;
+    info->srcCount = 0;
+
+    if (op1 != nullptr)
     {
-        info->srcCount += GetOperandInfo(intrinsicTree->gtOp.gtOp2);
+        if (op1->OperIsList())
+        {
+            for (GenTreeArgList* list = op1->AsArgList(); list != nullptr; list = list->Rest())
+            {
+                info->srcCount += GetOperandInfo(list->Current());
+            }
+        }
+        else
+        {
+            info->srcCount += GetOperandInfo(op1);
+        }
     }
 
-#ifdef _TARGET_X86_
-    if (intrinsicTree->gtHWIntrinsicId == NI_SSE42_Crc32)
+    if (op2 != nullptr)
     {
-        // CRC32 may operate over "byte" but on x86 only RBM_BYTE_REGS can be used as byte registers.
-        //
-        // TODO - currently we use the BaseType to bring the type of the second argument
-        // to the code generator. May encode the overload info in other way.
-        var_types srcType = intrinsicTree->gtSIMDBaseType;
-        if (varTypeIsByte(srcType))
+        info->srcCount += GetOperandInfo(op2);
+    }
+
+    switch (intrinsicID)
+    {
+        case NI_SSE41_BlendVariable:
         {
-            LocationInfoListNode* op2Info = useList.GetSecond(INDEBUG(intrinsicTree->gtGetOp2()));
-            op2Info->info.setSrcCandidates(this, RBM_BYTE_REGS);
+            if (!compiler->canUseVexEncoding())
+            {
+                // SSE4.1 blendv* hardcode the mask vector (op3) in XMM0
+                LocationInfoListNode* op3Info = useList.Begin()->Next()->Next();
+                LocationInfoListNode* op2Info = useList.Begin()->Next();
+                op2Info->info.isDelayFree = true;
+                op3Info->info.setSrcCandidates(this, RBM_XMM0);
+                op3Info->info.isDelayFree = true;
+                info->hasDelayFreeSrc = true;
```

Comment:
@mikedn Now this works fine on op1 and op2, but op3 is still not bounded to XMM0.
```csharp
var vf1 = Unsafe.Read<Vector128<float>>(ptr1);
var vf2 = Unsafe.Read<Vector128<float>>(ptr2);
var vf3 = Unsafe.Read<Vector128<float>>(ptr3);
var vf4 = Sse.Add(vf1, vf2);
var vf5 = Sse41.BlendVariable(vf1, vf3, vf2);
var vf6 = Sse41.BlendVariable(vf1, vf5, vf4);
var vf7 = Sse41.BlendVariable(vf1, vf2, vf6);
```
```asm
...
IN0053: 0001E1 movups   xmm0, xmmword ptr [rax]
IN0054: 0001E4 movups   xmm1, xmmword ptr [rdi+16]
IN0055: 0001E8 movaps   xmm2, xmm6
IN0056: 0001EB addps    xmm2, xmm0
IN0057: 0001EE movaps   xmmword ptr [V05 rbp-A0H], xmm0
IN0058: 0001F5 movaps   xmm3, xmm6
IN0059: 0001F8 blendvps  xmm3, xmm1
IN005a: 0001FD movaps   xmm1, xmm6
IN005b: 000200 blendvps  xmm1, xmm3
IN005c: 000205 movaps   xmm2, xmmword ptr [V05 rbp-A0H]
IN005d: 00020C blendvps  xmm6, xmm2
```

Reply:
@fiigii Could you keep commits separate for this problematic part of code just to track what and how was working or not working? Perhaps it could help to dig faster into problem.

Reply:
@4creators okay, I will try to make the commits clearer at next time 😄

Reply:
> Now this works fine on op1 and op2, but op3 is still not bounded to XMM0

Yes, you need to emit a `movaps xmm0, op3's reg` in codegen (if op3's reg is not already xmm0). The `gtRegNum` of a node contains the register where that node has stored the value. That may or may not be the same register that's needed at the "use" position.

Reply:
> I will try to make the commits clearer at next time

That's a waste of time. If you have a problem then explain the problem and if possible point at the relevant code. It makes no difference if you have 1 or 10 commits.